File: leetcode/leetcode2685.cpp
#include<vector>
#include<utility>
using namespace std;
class Solution {
public:
    vector<vector<int>> g;
    vector<bool> visited;
    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        g=vector<vector<int>>(n);
        visited=vector<bool>(n);
        int ans=0;
        for (int i = 0; i < edges.size(); i++)
        {
            int a=edges[i][0];
            int b=edges[i][1];
            g[a].push_back(b);
            g[b].push_back(a);
        }
        for (int i=0;i<n;i++){
            if(!visited[i]){
                pair<int,int> res=dfs(i);
                if (res.second==res.first*(res.first-1)){
                    ans++;
                }
            }
        }
        return ans;
    }
    pair<int,int> dfs(int x){
        visited[x]=true;
        pair<int,int> res;
        res.first=1;
        res.second=g[x].size();
        for (int i=0;i<g[x].size();i++){
            if(!visited[g[x][i]]){
                pair<int,int> count=dfs(g[x][i]);
                res.first+=count.first;
                res.second+=count.second;
            }
        }
        return res;
    }
};

File: leetcode/leetcode2685.py
#https://leetcode.cn/problems/count-the-number-of-complete-components/?envType=daily-question&envId=2026-07-11
from typing import List
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        ans=0
        g=[[] for _ in range(n)]
        for a,b in edges:
            g[a].append(b)
            g[b].append(a)
        visit=[False for _ in range(n)]
        def dfs(x):
            visit[x]=True
            node_count = 1
            degree_sum = len(g[x])
            for y in g[x]:
                if not visit[y]:
                    sub_count, sub_degree = dfs(y)
                    node_count += sub_count
                    degree_sum += sub_degree
            return node_count, degree_sum
        for x in range(n):
            if not visit[x]:
                node_count, degree_sum = dfs(x)
                if degree_sum == node_count * (node_count - 1):
                    ans += 1
        return ans


